tests/storage/test_event_federation.py

```diff
@@ -571,11 +587,449 @@ def prev_event_format(prev_event_id: str) -> Union[Tuple[str, dict], str]:
+
+        # Try at "B"
+        backfill_points = self.get_success(
+            self.store.get_backfill_points_in_room(room_id, depth_map["B"])
+        )
+        backfill_event_ids = [backfill_point[0] for backfill_point in backfill_points]
+        # Only the backfill points that we didn't record earlier exist here.
+        self.assertListEqual(backfill_event_ids, ["b6", "2", "b1"])
+
+    def test_get_backfill_points_in_room_attempted_event_retry_after_backoff_duration(
+        self,
+    ):
+        """
+        Test to make sure after we fake attempt to backfill event "b3" many times,
+        we can see retry and see the "b3" again after the backoff timeout duration
+        has exceeded.
+        """
+        setup_info = self._setup_room_for_backfill_tests()
+        room_id = setup_info.room_id
+        depth_map = setup_info.depth_map
+
+        # Record some attempts to backfill these events which will make
+        # `get_backfill_points_in_room` exclude them because we
+        # haven't passed the backoff interval.
+        self.get_success(self.store.record_event_failed_backfill_attempt(room_id, "b3"))
+        self.get_success(self.store.record_event_failed_backfill_attempt(room_id, "b1"))
+        self.get_success(self.store.record_event_failed_backfill_attempt(room_id, "b1"))
+        self.get_success(self.store.record_event_failed_backfill_attempt(room_id, "b1"))
+        self.get_success(self.store.record_event_failed_backfill_attempt(room_id, "b1"))
+
+        # Now advance time by 2 hours and we should only be able to see "b3"
+        # because we have waited long enough for the single attempt (2^1 hours)
+        # but we still shouldn't see "b1" because we haven't waited long enough
+        # for this many attempts. We didn't do anything to "b2" so it should be
+        # visible regardless.
+        self.reactor.advance(datetime.timedelta(hours=2).total_seconds())
+
+        # Try at "A" and make sure that "b3" is not in the list because we've
+        # already attemted many times
+        backfill_points = self.get_success(
+            self.store.get_backfill_points_in_room(room_id, depth_map["A"])
+        )
+        backfill_event_ids = [backfill_point[0] for backfill_point in backfill_points]
+        self.assertListEqual(backfill_event_ids, ["b3", "b2"])
+
+        # Now advance time by 20 hours (above 2^4 because we made 4 attemps) and
+        # see if we can now backfill it
+        self.reactor.advance(datetime.timedelta(hours=20).total_seconds())
+
+        # Try at "A" again after we advanced enough time and we should see "b3" again
+        backfill_points = self.get_success(
+            self.store.get_backfill_points_in_room(room_id, depth_map["A"])
+        )
+        backfill_event_ids = [backfill_point[0] for backfill_point in backfill_points]
+        self.assertListEqual(backfill_event_ids, ["b3", "b2", "b1"])
+
+    def _setup_room_for_insertion_backfill_tests(self) -> _BackfillSetupInfo:
```

Comment:
The next set of 3 tests for `get_insertion_event_backward_extremities_in_room` are very similar to the 3 tests added for `get_backfill_points_in_room`.

We could probably re-use and generalize these into a test matrix but the everything would be less straightforward. I'm leaning towards leaving this as-is vs complicating it.